### ISSUE TITLE
Investigate memory buffer and class handling in CIFAR50Incremental experiment

### ISSUE THREAD

**Issue by SamanNaeem** [SamanNaeem/Continual-Knowledge-Distillation-for-Class-Incremental-Learning-on-CIFAR-IOO#1]
`# ===================== Cell 1: Imports & Config =====================
import torch
import torch.nn as nn
import torch.optim as optim
import torchvision
from torchvision import datasets, transforms
from torch.utils.data import DataLoader, Dataset, Subset
import numpy as np
from tqdm import tqdm
import random
import matplotlib.pyplot as plt
import seaborn as sns
from collections import defaultdict
import torch.nn.functional as F

# ========== Editable Experiment Config ==========
num_tasks = 5           # Number of tasks to split (50 classes total)
buffer_size = 2000      # Total replay / exemplar memory size
memory_batch_size = 64  # Number of replay samples per training batch (fix for NaNs)
ewc_lambda = 2500       # Regularizer strength (EWC or MAS)
use_icarl = False       # Use iCaRL memory instead of vanilla replay buffer
regularizer = 'ewc'     # 'ewc', 'mas', or None
num_epochs = 25         # Epochs per task
batch_size = 64         # Training batch size
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

# For reproducibility
torch.manual_seed(42)
np.random.seed(42)
random.seed(42)

print(f"Using device: {device}")
`

**Comment by SamanNaeem** [comment]
`# ===================== Cell 2: Dataset, Data Utilities =====================
class TransformedSubset(Dataset):
    def __init__(self, dataset, indices, transform=None):
        self.dataset = dataset
        self.indices = indices
        self.transform = transform

    def __getitem__(self, idx):
        x, y = self.dataset[self.indices[idx]]
        return (self.transform(x) if self.transform else x), y

    def __len__(self):
        return len(self.indices)

class CIFAR50Incremental:
    def __init__(self, root='./data', num_tasks=5, initial_shuffle=True):
        self.root = root
        self.num_tasks = num_tasks
        self.classes_per_task = 50 // num_tasks
        self.total_classes = 50

        self.train_transform = transforms.Compose([
            transforms.RandomCrop(32, padding=4),
            transforms.RandomHorizontalFlip(),
            transforms.ToTensor(),
            transforms.Normalize((0.5071, 0.4867, 0.4408), (0.2675, 0.2565, 0.2761))
        ])
        self.test_transform = transforms.Compose([
            transforms.ToTensor(),
            transforms.Normalize((0.5071, 0.4867, 0.4408), (0.2675, 0.2565, 0.2761))
        ])

        full_train = datasets.CIFAR100(root=root, train=True, download=True)
        full_test = datasets.CIFAR100(root=root, train=False, download=True)
        selected_classes = np.arange(50)
        train_idx = [i for i, (_, y) in enumerate(full_train) if y in selected_classes]
        test_idx = [i for i, (_, y) in enumerate(full_test) if y in selected_classes]

        self.base_train = TransformedSubset(full_train, train_idx)
        self.base_test = TransformedSubset(full_test, test_idx)

        self.class_order = np.arange(50)
        if initial_shuffle:
            np.random.shuffle(self.class_order)

    def get_task_data(self, task_id):
        assert 0 <= task_id < self.num_tasks
        start = task_id * self.classes_per_task
        end = (task_id + 1) * self.classes_per_task
        task_classes = self.class_order[start:end]

        train_mask = np.isin(np.array([y for _, y in self.base_train]), task_classes)
        test_mask = np.isin(np.array([y for _, y in self.base_test]), task_classes)

        train_subset = TransformedSubset(self.base_train, np.where(train_mask)[0], self.train_transform)
        test_subset = TransformedSubset(self.base_test, np.where(test_mask)[0], self.test_transform)
        return train_subset, test_subset, task_classes
`

**Comment by SamanNaeem** [comment]
`# ===================== Cell 3: Model (ResNet18 w/ Attention) =====================
class AttentionLayer(nn.Module):
    def __init__(self, features):
        super().__init__()
        self.attention = nn.Sequential(
            nn.Linear(features, features//4),
            nn.ReLU(),
            nn.Linear(features//4, features),
            nn.Sigmoid()
        )
    def forward(self, x):
        attn_weights = self.attention(x.mean(dim=[2,3]))
        return x * attn_weights.unsqueeze(-1).unsqueeze(-1)

class AttnBasicBlock(nn.Module):
    expansion = 1
    def __init__(self, in_planes, planes, stride=1):
        super().__init__()
        self.conv1 = nn.Conv2d(in_planes, planes, kernel_size=3, stride=stride, padding=1, bias=False)
        self.bn1 = nn.BatchNorm2d(planes)
        self.conv2 = nn.Conv2d(planes, planes, kernel_size=3, stride=1, padding=1, bias=False)
        self.bn2 = nn.BatchNorm2d(planes)
        self.attn = AttentionLayer(planes * self.expansion)
        self.shortcut = nn.Sequential()
        if stride != 1 or in_planes != self.expansion*planes:
            self.shortcut = nn.Sequential(
                nn.Conv2d(in_planes, self.expansion*planes, kernel_size=1, stride=stride, bias=False),
                nn.BatchNorm2d(self.expansion*planes)
            )
    def forward(self, x):
        out = F.relu(self.bn1(self.conv1(x)))
        out = self.bn2(self.conv2(out))
        out += self.shortcut(x)
        out = F.relu(out)
        return self.attn(out)

class ResNet(nn.Module):
    def __init__(self, block, num_blocks, num_classes=50):
        super().__init__()
        self.in_planes = 64
        self.conv1 = nn.Conv2d(3, 64, kernel_size=3, stride=1, padding=1, bias=False)
        self.bn1 = nn.BatchNorm2d(64)
        self.layer1 = self._make_layer(block, 64, num_blocks[0], stride=1)
        self.layer2 = self._make_layer(block, 128, num_blocks[1], stride=2)
        self.layer3 = self._make_layer(block, 256, num_blocks[2], stride=2)
        self.layer4 = self._make_layer(block, 512, num_blocks[3], stride=2)
        self.linear = nn.Linear(512*block.expansion, num_classes)
    def _make_layer(self, block, planes, num_blocks, stride):
        strides = [stride] + [1]*(num_blocks-1)
        layers = []
        for stride in strides:
            layers.append(block(self.in_planes, planes, stride))
            self.in_planes = planes * block.expansion
        return nn.Sequential(*layers)
    def forward(self, x):
        out = F.relu(self.bn1(self.conv1(x)))
        out = self.layer1(out)
        out = self.layer2(out)
        out = self.layer3(out)
        out = self.layer4(out)
        out = F.avg_pool2d(out, 4)
        out = out.view(out.size(0), -1)
        out = self.linear(out)
        return out

def ResNet18(num_classes=50):
    return ResNet(AttnBasicBlock, [2, 2, 2, 2], num_classes)
`

**Comment by SamanNaeem** [comment]
`# ===================== Cell 4: Replay & Exemplar Memory =====================
class ReplayBuffer:
    def __init__(self, buffer_size=2000):
        self.buffer_size = buffer_size
        self.images, self.labels = [], []
    def add(self, x, y):
        for xi, yi in zip(x, y):
            if len(self.images) < self.buffer_size:
                self.images.append(xi.cpu() if xi.is_cuda else xi)
                self.labels.append(int(yi))
            else:
                idx = np.random.randint(0, self.buffer_size)
                self.images[idx] = xi.cpu() if xi.is_cuda else xi
                self.labels[idx] = int(yi)
    def sample(self, n):
        if len(self.images) == 0:
            return [], []
        idxs = np.random.choice(len(self.images), n, replace=len(self.images)<n)
        xs = torch.stack([self.images[i] for i in idxs])
        ys = torch.tensor([self.labels[i] for i in idxs])
        return xs, ys

class iCaRLExemplarMemory:
    def __init__(self, memory_size, feature_extractor):
        self.memory_size = memory_size
        self.feature_extractor = feature_extractor
        self.exemplar_sets = defaultdict(list)
        self.class_order = []
    def construct_exemplar_set(self, loader, class_label, m):
        X = []
        for x, y in loader:
            if y[0] == class_label:
                X.append(x.to(device))
        if len(X) == 0: return
        feats = torch.cat([self.feature_extractor(x.unsqueeze(0)) for x in X], 0)
        class_mean = feats.mean(0)
        selected, mean_so_far = [], torch.zeros_like(class_mean)
        for _ in range(m):
            remain = [i for i in range(len(X)) if i not in selected]
            best_i, min_dist = None, float('inf')
            for i in remain:
                candidate = (mean_so_far + feats[i]) / (len(selected) + 1)
                dist = torch.norm(class_mean - candidate)
                if dist < min_dist:
                    best_i, min_dist = i, dist
            selected.append(best_i)
            mean_so_far += feats[best_i]
        self.exemplar_sets[class_label] = [X[i].cpu() for i in selected]
    def reduce(self, new_total):
        per_class = max(1, new_total // max(1, len(self.exemplar_sets)))
        for cls in self.exemplar_sets:
            self.exemplar_sets[cls] = self.exemplar_sets[cls][:per_class]
    def get_all(self, n=memory_batch_size):
        exemplars, labels = [], []
        for cls in self.exemplar_sets:
            exemplars += self.exemplar_sets[cls]
            labels += [cls]*len(self.exemplar_sets[cls])
        if len(exemplars) == 0:
            return [], []
        idxs = np.random.choice(len(exemplars), min(n, len(exemplars)), replace=False)
        ex = [exemplars[i] for i in idxs]
        ey = [labels[i] for i in idxs]
        return ex, ey
`

**Comment by SamanNaeem** [comment]
`# ===================== Cell 5: Regularizers: EWC, MAS =====================
class EnhancedEWC:
    def __init__(self, model, dataloader, device):
        self.model = model
        self.device = device
        self.importance = {}
        self.params = {n: p.detach().clone() for n, p in model.named_parameters() if p.requires_grad}
        self.compute_fisher(dataloader)
    def compute_fisher(self, dataloader):
        fisher = {n: torch.zeros_like(p) for n,p in self.model.named_parameters() if p.requires_grad}
        self.model.eval()
        for inputs, targets in tqdm(dataloader, desc="EWC Fisher"):
            inputs,targets = inputs.to(self.device), targets.to(self.device)
            self.model.zero_grad()
            outputs = self.model(inputs)
            loss = F.cross_entropy(outputs, targets)
            loss.backward()
            for n,p in self.model.named_parameters():
                if p.grad is not None: fisher[n] += p.grad.data.pow(2) / len(dataloader)
        self.importance = fisher
    def penalty(self, model, lambda_=ewc_lambda):
        loss = 0
        for n, p in model.named_parameters():
            if n in self.importance:
                loss += (self.importance[n] * (p - self.params[n]).pow(2)).sum()
        return lambda_ * loss

class MAS:
    def __init__(self, model, dataloader, device):
        self.model = model
        self.device = device
        self.importance = self.compute_importance(dataloader)
        self.params = {n: p.detach().clone() for n,p in model.named_parameters() if p.requires_grad}
    def compute_importance(self, dataloader):
        importance = {n: torch.zeros_like(p) for n,p in self.model.named_parameters() if p.requires_grad}
        self.model.eval()
        for x, _ in tqdm(dataloader, desc="MAS importance"):
            x = x.to(self.device)
            self.model.zero_grad()
            outputs = self.model(x)
            outputs.abs().sum().backward()
            for n,p in self.model.named_parameters():
                if p.grad is not None: importance[n] += p.grad.abs() / len(dataloader)
        return importance
    def penalty(self, model, lambda_=ewc_lambda):
        loss = 0
        for n, p in model.named_parameters():
            loss += (self.importance[n] * (p-self.params[n]).pow(2)).sum()
        return lambda_ * loss
`

**Comment by SamanNaeem** [comment]
`# ===================== Cell 6: Training/Eval Functions (Fixed Memory Sampling + Debug) =====================
def evaluate(model, loader):
    model.eval()
    correct, total = 0, 0
    with torch.no_grad():
        for x, y in loader:
            x, y = x.to(device), y.to(device)
            out = model(x)
            pred = out.argmax(1)
            correct += (pred == y).sum().item()
            total += y.size(0)
    return 100.*correct/total if total > 0 else 0.0

def train_task(student, optimizer, loader, memory, teacher, reg, task_id, num_tasks):
    student.train()
    for epoch in range(num_epochs):
        running_loss, correct, total = 0., 0, 0
        total_batches = 0
        for x, y in tqdm(loader, desc=f'Task {task_id+1} Epoch {epoch+1}'):
            x, y = x.to(device), y.to(device)
            # Sample limited replay memory for this batch
            if memory is not None and (hasattr(memory, 'get_all') or hasattr(memory, 'sample')):
                if use_icarl and hasattr(memory, 'get_all'):
                    ex, ey = memory.get_all(n=memory_batch_size)
                    if ex:
                        ex = torch.stack(ex).to(device)
                        ey = torch.tensor(ey).to(device)
                        x = torch.cat([x, ex], dim=0)
                        y = torch.cat([y, ey], dim=0)
                elif hasattr(memory, 'sample'):
                    ex, ey = memory.sample(memory_batch_size)
                    if len(ex):
                        x = torch.cat([x, ex.to(device)], dim=0)
                        y = torch.cat([y, ey.to(device)], dim=0)

            # Check label range
            if y.max() >= student.linear.out_features:
                print(f"Label out of range ({y.max().item()}) for classifier of size {student.linear.out_features}! Skipping batch.")
                continue
            # Check NaNs in input/output
            with torch.no_grad():
                if torch.isnan(x).any() or torch.isnan(y).any():
                    print("NaN detected in inputs or targets. Skipping batch.")
                    continue
            out = student(x)
            if torch.isnan(out).any():
                print("NaN detected in outputs. Skipping batch.")
                continue

            ce_loss = F.cross_entropy(out, y)
            kd_loss = 0
            if teacher is not None:
                with torch.no_grad():
                    t_logits = teacher(x)
                T = 2
                kd_loss = F.kl_div(F.log_softmax(out/T, dim=1), F.softmax(t_logits/T, dim=1), reduction='batchmean') * T*T

            loss = 0.5 * ce_loss + 0.5 * kd_loss
            # if reg is not None:
            #     loss += reg.penalty(student, lambda_=ewc_lambda)

            optimizer.zero_grad()
            loss.backward()
            optimizer.step()

            running_loss += loss.item()
            _, pred = out.max(1)
            total += y.size(0)
            correct += (pred == y).sum().item()
            total_batches += 1

        if total_batches == 0:
            print(f"Epoch {epoch+1}: No valid batches run!")
        else:
            print(f"Epoch {epoch+1}: Loss={running_loss/total_batches:.4f}, Acc={100.*correct/total:.2f}%")
`

**Comment by SamanNaeem** [comment]
`# ===================== Cell 7: Main Loop & Plot =====================
def main():
    dataset = CIFAR50Incremental(num_tasks=num_tasks)
    student = ResNet18(num_classes=50).to(device)
    teacher = None

    if use_icarl:
        # Feature extractor is all layers except the final linear
        feature_extractor = nn.Sequential(*(list(student.children())[:-1])).to(device)
        memory = iCaRLExemplarMemory(memory_size=buffer_size, feature_extractor=feature_extractor)
    else:
        memory = ReplayBuffer(buffer_size=buffer_size)

    acc_matrix = np.zeros((num_tasks, num_tasks))
    max_acc = [0]*num_tasks

    for task_id in range(num_tasks):
        print(f"\n=== Task {task_id+1}/{num_tasks} ===")
        train_set, test_set, task_classes = dataset.get_task_data(task_id)
        tr_loader = DataLoader(train_set, batch_size=batch_size, shuffle=True)
        te_loader = DataLoader(test_set, batch_size=batch_size, shuffle=False)

        # Populate memory buffer / iCaRL exemplars
        if use_icarl:
            student.eval()
            for c in task_classes:
                cls_loader = DataLoader([d for d in train_set if d[1] == c], batch_size=1)
                memory.construct_exemplar_set(cls_loader, int(c), max(1, buffer_size // ((task_id + 1)*len(task_classes))))
            memory.reduce(buffer_size)
        else:
            for i in range(min(buffer_size // num_tasks, len(train_set))):
                xi, yi = train_set[i]
                memory.add([xi], [yi])

        optimizer = optim.SGD(student.parameters(), lr=0.05, momentum=0.9, weight_decay=5e-4)
        reg = None
        if regularizer == 'ewc' and task_id > 0:
            reg = EnhancedEWC(student, tr_loader, device)
        elif regularizer == 'mas' and task_id > 0:
            reg = MAS(student, tr_loader, device)

        # Set teacher for KD
        teacher = ResNet18(num_classes=50).to(device)
        teacher.load_state_dict(student.state_dict())
        teacher.eval()

        train_task(student, optimizer, tr_loader, memory, teacher, reg, task_id, num_tasks)

        # Evaluate on all seen tasks so far
        for eval_tid in range(task_id+1):
            _, eval_test_set, _ = dataset.get_task_data(eval_tid)
            eval_loader = DataLoader(eval_test_set, batch_size=batch_size, shuffle=False)
            acc = evaluate(student, eval_loader)
            acc_matrix[task_id, eval_tid] = acc
            max_acc[eval_tid] = max(max_acc[eval_tid], acc)
            print(f"Accuracy on Task {eval_tid+1}: {acc:.2f}%")

    forgetting = [max_acc[i] - acc_matrix[-1, i] for i in range(num_tasks)]

    plt.figure(figsize=(15, 5))
    plt.subplot(1, 2, 1)
    sns.heatmap(acc_matrix, annot=True, fmt=".1f", cmap="YlGnBu",
                xticklabels=[f"T{i+1}" for i in range(num_tasks)],
                yticklabels=[f"After T{i+1}" for i in range(num_tasks)])
    plt.title("Accuracy matrix")

    plt.subplot(1, 2, 2)
    plt.plot(range(1, num_tasks + 1), forgetting, marker='o')
    plt.xlabel("Task")
    plt.ylabel("Forgetting (%)")
    plt.title("Forgetting Over Tasks")
    plt.grid(True)
    plt.tight_layout()
    plt.show()

if __name__ == "__main__":
    # Run the full experiment
    # Debug: check dataset and labels before training
    dataset_check = CIFAR50Incremental(num_tasks=5)
    train_data, test_data, classes = dataset_check.get_task_data(0)
    print(f"Debug - Task 0 classes: {classes}")
    print(f"Debug - Train samples: {len(train_data)}, Test samples: {len(test_data)}")
    for x, y in DataLoader(train_data, batch_size=10):
        print(f"Debug - Sample labels: {y}")
        break
    # Now run training and evaluation
    main()
`
